cluster-autoscaler/simulator/predicatechecker/schedulerbased.go

```diff
@@ -181,6 +212,41 @@ func (p *SchedulerBasedPredicateChecker) CheckPredicates(clusterSnapshot cluster
 			filterReasons,
 			p.buildDebugInfo(filterName, nodeInfo))
 	}
+
+	return p.runExtendersForNode(pod, nodeInfo)
+}
+
+// runExtendersForNode runs `Filter` function of all extenders for a particular pod on a single node
+func (p *SchedulerBasedPredicateChecker) runExtendersForNode(pod *apiv1.Pod, nodeInfo *schedulerframework.NodeInfo) *PredicateError {
+	for _, extender := range p.framework.Extenders() {
+		if extender.IsInterested(pod) {
+			filteredNodes, failedNodesMap, unresolvableNodesMap, err := extender.Filter(pod, []*apiv1.Node{nodeInfo.Node()})
```

Comment:
nit: Looks like `extender.Filter` accepts a list of nodes - wouldn't it make more sense to run this once after `PreFilter`, but before `Filter` and pass all the nodes that came out of `PreFilter`?

Reply:
Passing all nodes `extender.Filter` at once so that we don't call it multiple times makes sense to me but same could be said of running filter plugins (even though cost of calling in-tree filter plugins per node won't be as high). I just took a safe bet here and went with the same approach as running filter plugins (might not be the best approach as you rightfully pointed out). I guess the question here would be, why didn't we do the same for running filter plugins?

`extender.Filter` is called in the upstream scheduler code after all `PreFilter` and `Filter` plugins are called. I don't see much benefit in changing the order. 
```go
File: pkg/scheduler/schedule_one.go
385: // Filters the nodes to find the ones that fit the pod based on the framework
386: // filter plugins and filter extenders.
387: func (sched *Scheduler) findNodesThatFitPod(ctx context.Context, fwk framework.Framework, state *framework.CycleState, pod *v1.Pod) ([]*v1.Node, framework.Diagnosis, error) {
... 
393: 	allNodes, err := sched.nodeInfoSnapshot.NodeInfos().List()
394: 	if err != nil {
395: 		return nil, diagnosis, err
396: 	}
397: 	// Run "prefilter" plugins.
398: 	preRes, s := fwk.RunPreFilterPlugins(ctx, state, pod)
399: 	...
413: 
414: 	...
438: 	feasibleNodes, err := sched.findNodesThatPassFilters(ctx, fwk, state, pod, diagnosis, nodes)
...
446: 
447: 	feasibleNodes, err = findNodesThatPassExtenders(sched.Extenders, pod, feasibleNodes, diagnosis.NodeToStatusMap)
448: 	if err != nil {
449: 		return nil, diagnosis, err
450: 	}

```
https://github.com/kubernetes/kubernetes/blob/8d18ae6fc245deb68534cab608056c5aae9160bd/pkg/scheduler/schedule_one.go#L385-L452

```go
File: pkg/scheduler/schedule_one.go
576: func findNodesThatPassExtenders(extenders []framework.Extender, pod *v1.Pod, feasibleNodes []*v1.Node, statuses framework.NodeToStatusMap) ([]*v1.Node, error) {
577: 	// Extenders are called sequentially.
578: 	// Nodes in original feasibleNodes can be excluded in one extender, and pass on to the next
579: 	// extender in a decreasing manner.
...
587: 
588: 		// Status of failed nodes in failedAndUnresolvableMap will be added or overwritten in <statuses>,
589: 		// so that the scheduler framework can respect the UnschedulableAndUnresolvable status for
590: 		// particular nodes, and this may eventually improve preemption efficiency.
591: 		// Note: users are recommended to configure the extenders that may return UnschedulableAndUnresolvable
592: 		// status ahead of others.
593: 		feasibleList, failedMap, failedAndUnresolvableMap, err := extender.Filter(pod, feasibleNodes)
...

```
https://github.com/kubernetes/kubernetes/blob/8d18ae6fc245deb68534cab608056c5aae9160bd/pkg/scheduler/schedule_one.go#L576-L627